### PR TITLE
Adds affine transformations

### LINKED ISSUE #1: Implement affine image transformation
One of the key features of GalSim is the ability to apply affine transformations to a GSObject, this is in particular needed for adding shear to a galaxy image.

In GalSim, these things are defined here for instance: https://github.com/GalSim-developers/GalSim/blob/013292ad45b6ef715acf70976e94fe21f624ee4c/galsim/transform.py#L98

And you can see here what happens when a GSObject is sheared: https://github.com/GalSim-developers/GalSim/blob/013292ad45b6ef715acf70976e94fe21f624ee4c/galsim/gsobject.py#L863
What happens is that an Affine Transformation is defined, and applied on the image.

The goal of this issue is to discuss the implementation of similar transformations within GalFlow using TensorFlow, so that the transformation can be differentiated with respect to the input Jacobian.
I've found these image tools in TensorFlow Addons: https://www.tensorflow.org/addons/api_docs/python/tfa/image/transform 
And I think this may be exactly what we need.

### PR DESCRIPTION
**Pull request by EiffL**
This PR aims to close #1 and adds the tools needed to implement affine transformations in TensorFlow, and to generate the transformation needed for shear. I've updated the [following notebook](https://github.com/DifferentiableUniverseInitiative/GalFlow/blob/affine-transformation/notebooks/Shear_transformation.ipynb) with an example of how to use the new API. It essentially looks like this:

```
import galflow as gf
import tensorflow as tf

# It needs a batch and channel dimension
image = tf.zeros([1,256,256,1])
g1 = tf.zeros([1]) # Same batch dimension as the image
g2 = tf.zeros([1])

# And compute the sheared image
sheared_image = gf.shear(image, g1, g2)
```
image

Internally, I've added a few tools to handle generic affine transformations, this is going to be useful at some point when implementing WCS. I've used a matrix representation that allows for easy combination of different transformations into a new one.

### CODE REVIEW

**Review by b-remy — COMMENTED** [review]
Looks very good! I think we don't need the `tfa-nightly` dependency unless for notebook examples, so maybe we don't need to specify it in the requirements. What do you think?

**Comment by EiffL** [comment]
Ok so I checked :-) and it doesnt work if I remove tfa-nightly from the setup.py, because it's needed for tfg, and tfg doesnt install it.

**Comment by EiffL** [comment]
But, you have a point there :-) We need to have some tests to make sure things install and work correctly I'm gonna add one

File: galflow/python/shear.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+__all__ = ["shear", "shear_transformation"]
+
+def shear_transformation(g1, g2, Fourier=False):
+  """
+  Function to compute the affine transformation corresponding to a given shear.
+  This function uses the reduced shear definition:
+
+    :math:`|g| = \frac{a - b}{a + b}`
+
+  TODO: better documentation (the following is stolen from galsim)
```
Comment (EiffL) [inline galflow/python/shear.py]: Fix documentation if it is a TODO

File: galflow/python/shear.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  Parameters
+  ----------
+    g1: Tensor
+      The first component of the shear in the "reduced shear" definition.
+    g2: Tensor
+     The second component of the shear in the "reduced shear" definition.
+
+  Returns
+  -------
+    Transformation matrix format, i.e [(batch), 3, 3]
```
Comment (EiffL) [inline galflow/python/shear.py:42]: use proper docstring formatting

File: galflow/python/shear.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  if Fourier:
+    jac = tf.transpose(jac, [0,2,1])
+  else:
+    jac = tf.linalg.inv(jac)
+  jac = tf.pad(jac, tf.constant([[0, 0], [0, 1],[0,1]]) )
+  jac = jac + tf.pad(tf.reshape(tf.ones_like(g1), [-1,1,1]), tf.constant([[0,0],[2,0],[2,0]]))
+  return jac
+
+def shear(img, g1, g2):
+  """ Convenience function to apply a shear to an input image or kimage
```
Comment (EiffL) [inline galflow/python/shear.py]: Add proper documentation

File: galflow/python/transform.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    assert transforms, "transforms cannot be empty"
+    with tf.name_scope(name or "compose_transforms"):
+        composed = transforms[0]
+        for tr in transforms[1:]:
+            # Multiply batches of matrices.
+            composed = tf.matmul(composed, tr)
+        return composed
+
+def shift_transformation(nx, ny):
+  """ Computes affine transformation for specified pixel shift.
```
Comment (EiffL) [inline galflow/python/transform.py:34]: Add proper documentation

File: galflow/python/transform.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+  ny = tf.convert_to_tensor(ny, dtype=tf.float32)
+  # Convert to transformation format
+  shift = tf.reshape(tf.concat([nx, ny], axis=1), [-1, 2, 1])
+  return tf.pad(shift,[[0,0],[0,1],[2,0]]) + tf.expand_dims(tf.eye(3),0)
+
+def transform(img, transform_matrix, keep_center=True):
+  """
+  Function for transforming an image or kimage.
+
+  TODO: improve documentation, doesnt handle chromaticity yet
```
Comment (EiffL) [inline galflow/python/transform.py]: Fix documentation TODO

**Review by EiffL — COMMENTED** [review]
I see that there is no test for the Fourier-domain interpolation, despite that functionality being exposed to the user, so we should add one

**Comment by EiffL** [comment]
@b-remy I reviewed my own code, just to show you that my PR was below standards on a number of points :-) I'll fix them and maybe ask you to do a second round of reviews.

File: galflow/python/shear.py (hunk 2 above)
Comment (EiffL) [reply galflow/python/shear.py:42]: I'm going to let it fly for now because we are going to change the documentation style anyway

**Comment by EiffL** [comment]
Me closing documentations comments by saying we will fix it later with a specific documentation issue
image

**Pull request merged**